src/3d/qgs3dexportobject.cpp

```diff
@@ -0,0 +1,159 @@
+ *   This program is free software; you can redistribute it and/or modify  *
+ *   it under the terms of the GNU General Public License as published by  *
+ *   the Free Software Foundation; either version 2 of the License, or     *
+ *   (at your option) any later version.                                   *
+ *                                                                         *
+ ***************************************************************************/
+
+#include "qgs3dexportobject.h"
+
+#include <QVector3D>
+#include <QDebug>
+#include <QDir>
+#include <QImage>
+
+Qgs3DExportObject::Qgs3DExportObject( const QString &name, const QString &parentName, QObject *parent )
+  : QObject( parent )
+  , mName( name )
+  , mParentName( parentName )
+  , mSmoothEdges( false )
+{
+}
+
+void Qgs3DExportObject::setupPositionCoordinates( const QVector<float> &positionsBuffer, float scale, const QVector3D translation )
+{
+  for ( int i = 0; i < positionsBuffer.size(); i += 3 )
+  {
+    for ( int j = 0; j < 3; ++j )
+    {
+      mVertxPosition << positionsBuffer[i + j] * scale + translation[j];
+    }
+  }
+
+  for ( int i = 0; i < positionsBuffer.size() / 3; ++i )
+  {
+    mIndexes << i + 1;
+  }
+}
+
+void Qgs3DExportObject::setupPositionCoordinates( const QVector<float> &positionsBuffer, const QVector<unsigned int> &faceIndex, float scale, const QVector3D translation )
+{
+  // TODO: delete vertices that are not used
+  for ( int i = 0; i < positionsBuffer.size(); i += 3 )
+  {
+    for ( int j = 0; j < 3; ++j )
+    {
+      mVertxPosition << positionsBuffer[i + j] * scale + translation[j];
+    }
+  }
+
+  for ( int i = 0; i < faceIndex.size(); i += 3 )
+  {
+    // skip invalid triangles
+    if ( faceIndex[i] == faceIndex[i + 1] && faceIndex[i + 1] == faceIndex[i + 2] ) continue;
+    for ( int j = 0; j < 3; ++j ) mIndexes << faceIndex[i + j] + 1;
+  }
+}
+
+void Qgs3DExportObject::setupNormalCoordinates( const QVector<float> &normalsBuffer )
+{
+  mNormals << normalsBuffer;
+}
+
+void Qgs3DExportObject::setupTextureCoordinates( const QVector<float> &texturesBuffer )
+{
+  mTexturesUV << texturesBuffer;
+}
+
+void Qgs3DExportObject::objectBounds( float &minX, float &minY, float &minZ, float &maxX, float &maxY, float &maxZ )
+{
+  for ( int vertice : mIndexes )
+  {
+    int heightIndex = ( vertice - 1 ) * 3 + 1;
+    minX = std::min( minX, mVertxPosition[heightIndex - 1] );
+    maxX = std::max( maxX, mVertxPosition[heightIndex - 1] );
+    minY = std::min( minY, mVertxPosition[heightIndex] );
+    maxY = std::max( maxY, mVertxPosition[heightIndex] );
+    minZ = std::min( minZ, mVertxPosition[heightIndex + 1] );
+    maxZ = std::max( maxZ, mVertxPosition[heightIndex + 1] );
+  }
+}
+
+void Qgs3DExportObject::saveTo( QTextStream &out, float scale, const QVector3D &center )
+{
+  // Set groups
+  // turns out grouping doest work as expected in blender
+
+  // smoothen edges
+  if ( mSmoothEdges ) out << "s on\n";
+  else out << "s off\n";
```

Comment:
code style: it's strongly preferred in qgis code to have if/else statements on separate lines:
```
if ( a )
  do_a();
else
  do_b();
```